File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys
import heapq
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

V, E = map(int, input().split())
parent = [x for x in range(V + 1)]
pq = []

# 간선 데이터를 최소 힙에 삽입
for _ in range(E):
    start, end, weight = map(int, input().split())
    heapq.heappush(pq, (weight, start, end))  # 가중치 기준 정렬

# 부모 노드를 찾는 함수
def find(a):
    if parent[a] == a:
        return a
    parent[a] = find(parent[a])  # 경로 압축
    return parent[a]

# 같은 집합인지 확인하는 함수
def check_same(a, b):
    a = find(a)
    b = find(b)
    return a == b

# 두 집합을 합치는 함수
def union(a, b):
    a = find(a)
    b = find(b)
    if a != b:
        if a < b:
            parent[b] = a
        else:
            parent[a] = b

total_sum = 0
edge_use = 0

# 크루스칼 알고리즘 실행
while edge_use < V - 1 and pq:
    weight, start, end = heapq.heappop(pq)  # 최소 가중치 간선 꺼내기
    if not check_same(start, end):  # 사이클이 생기지 않으면
        total_sum += weight
        union(start, end)
        edge_use += 1

print(total_sum)

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.StringTokenizer;

class Node implements Comparable<Node> {
    int start;
    int end;
    int weight;

    Node(int start, int end, int weight) {
        this.start = start;
        this.end = end;
        this.weight = weight;
    }

    @Override
    public int compareTo(Node o) {
        return weight - o.weight;
    }

}


public class Main {
    static int V, E;
    static int[] parent;
    static Queue<Node> pq;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine());

        V = Integer.parseInt(st.nextToken()); // 정점의 개수
        E = Integer.parseInt(st.nextToken()); // 간선의 개수

        parent = new int[V + 1];
        for (int i = 0; i <= V; i++) {
            parent[i] = i;
        }

        pq = new PriorityQueue<>();
        while (E-- > 0) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());
            pq.add(new Node(start, end, weight));
        }

        int totalSum = 0;
        while(!pq.isEmpty()) {
            Node curNode = pq.poll();
            int start = curNode.start;
            int end = curNode.end;
            int weight = curNode.weight;

            if (!checkSame(start, end)) {
                totalSum += weight;
                union(start, end);
            }

        }
        sb.append(totalSum);
        System.out.println(sb);
    }

    private static void union(int a, int b) {
        a = find(a);
        b = find(b);
        if (a != b) {
            if (a < b) parent[b] = a;
            else parent[a] = b;
        }
    }

    private static boolean checkSame(int a, int b) {
        a = find(a);
        b = find(b);
        if (a == b) {
            return true;
        }
        return false;
    }

    private static int find(int a) {
        if (parent[a] == a) return a;
        else return parent[a] = find(parent[a]);

    }

}